File: 프로그래머스/2/135807. 숫자 카드 나누기/숫자 카드 나누기.py
def solution(arrayA, arrayB):
    gcdA = arrayA[0]
    gcdB = arrayB[0]

    for num in arrayA[1:]:
        gcdA = calculate_gcd(num, gcdA)

    for num in arrayB[1:]:
        gcdB = calculate_gcd(num, gcdB)

    for num in arrayB:
        if num % gcdA == 0:
            gcdA = 0
            break

    for num in arrayA:
        if num % gcdB == 0:
            gcdB = 0
            break
            
    return max(gcdA, gcdB)

def calculate_gcd(a, b):
    while b:
        a, b = b, a % b
    return a

File: 프로그래머스/2/135807. 숫자 카드 나누기/숫자 카드 나누기.java
class Solution {
    public int solution(int[] arrayA, int[] arrayB) {
        int gcdA = arrayA[0];
        int gcdB = arrayB[0];

        for (int i = 1; i < arrayA.length; i++) {
            gcdA = calculateGCD(arrayA[i], gcdA);
        }

        for (int i = 1; i < arrayB.length; i++) {
            gcdB = calculateGCD(arrayB[i], gcdB);
        }

        for (int num : arrayB) {
            if (num % gcdA == 0) {
                gcdA = 0;
                break;
            }
        }

        for (int num : arrayA) {
            if (num % gcdB == 0) {
                gcdB = 0;
                break;
            }
        }

        return Math.max(gcdA, gcdB);
    }

    
    private static int calculateGCD(int a, int b) {
        if (b == 0) {
            return a;
        }
        return calculateGCD(b, a % b);
    }
}